# Project: Analyzing App User Numbers for the App Store and Google Play Market

## Introduction

Suppose we are working as a data analyst for a company that builds Android and iOS mobile apps. The company makes apps availiable both on Google Play and in the App Store. The company only buids apps that are **free** to download and install, and the main source of revenue consists of **in-app ads**.

This means that **the number of users of our apps** determines our revenue for any give app — the more users who see and engage with the ads, the better.

### Project Goal

***The goal of this project*** is to analyze data to help our developers understand what type of apps are likely to attract more users.

### Data Source

To achieve the goal, we'll need to collect and analyze data about mobile apps available on Google Play and the App Store.

As of 3rd quarter 2022, there were approximately 1.6 million iOS apps available on the App Store, and 3.6 million Android apps on Google Play (source: [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/)).

Collecting data for over 5 million apps requires a significant amount of time and money, so we'll try to analyze a sample of the data instead.

Currently, there are two existing datasets that seem to meet our goals:

- A dataset containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. From the link: https://dq-content.s3.amazonaws.com/350/googleplaystore.csv
- A dataset containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. From the link: https://dq-content.s3.amazonaws.com/350/AppleStore.csv

## 1. Opening and Exploring the Data

We will start by opening and exploring these two datasets.

### 1.1 Open the datasets

In [1]:
from csv import reader

### the AppleStore Dataset ###
opened_file=open('AppleStore.csv')
read_file=reader(opened_file)
ios=list(read_file)
ios_header=ios[0]
ios = ios[1:]

### the GooglePlayStore Dataset ###
opened_file=open('googleplaystore.csv')
read_file=reader(opened_file)
android=list(read_file)
android_header=android[0]
android = android[1:]

### 1.2 define the `explore_data()` function

To make it easier to explore the datasets, we created a function named `explore_data()` that we can repeatedly use to print rows in a readable way.

The function can print out entries of the dataset from specified start and end indexes. Also, it has an option to show the total number of rows and columns of a dataset.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

### 1.3 Explore the datasets

We'll then use the created function to explore the two datasets:

#### The App Store Dataset:

In [3]:
print("ios data")
print('\n')
print(ios_header)
print('\n')
explore_data(ios,0,5,True)

ios data


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


#### The Android Dataset:

In [4]:
print("android data")
print('\n')
print(android_header)
print('\n') 
explore_data(android,0,5,True)

android data


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'F

## 2. Data Cleaning

### 2.1 Deleting Wrong Data

We first check whether there are entries that contain missing value:

In [5]:
for row in ios:
    if len(row)!=len(ios_header):
        print(row)
        print('\n')
        print("ios Index postion is:", ios.index(row))
        
for row in android:
    if len(row)!=len(android_header):
        print(row)
        print('\n')
        print("Android Index postion is:", android.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Android Index postion is: 10472


From the output of the above cell, we can tell that the index 10472 row in the ios dataset contains a missing value. We therefore delete this row.

In [6]:
del android[10472]  # don't run this line more than once

### 2.2 Removing Duplicate Entries

#### 2.2.1 Check whether there are duplicate entries in the datasets

When we explore the Google P{lay dataset, we find that some apps have more than one entries.

For example, the application Instagram has four entries:

In [7]:
for app in android:
    name = app[0]
    if name =='Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [8]:
duplicate_apps_android = []
unique_apps_android = []     
        

for app in android:
    name = app[0]
    if name in unique_apps_android:
        duplicate_apps_android.append(name)
    else:
        unique_apps_android.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps_android))
print('\n')
print('Example of duplicate apps:', duplicate_apps_android[:15])

Number of duplicate apps: 1181


Example of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


From the above cell output, we can tell that there are 1180 duplicate apps in total.

Since we don't want to count certain apps more than once when we analyze data, we need to remove the duplicate entries and keep only one entry per app. 

One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

In the next step, we will set up a rule to remove the duplicate entries and keep only one entry per app:

#### 2.2.2 Set up a rule to remove the duplicate entries

Take Instagram as an example:

In [9]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


The duplicated entries have different number of reviews, indicating that the data was collected at different times. The highest number of reviews represent the most recent entry.

Therefore, we keep the entry with the highest reviews and delect all others.

#### 2.2.3 Create a dictionary containing unique app name and the highest number of reviews

In the next step, we try to create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app.

In [10]:
reviews_max={}

for row in android:
    name = row[0]
    n_reviews=float(row[3])
    
    if name in reviews_max and n_reviews > reviews_max[name]:
        reviews_max[name]=n_reviews
        
    elif name not in reviews_max:
        reviews_max[name]=n_reviews


#### 2.2.4 Use the dictionary to remove the duplicated rows

Then, we use the `reviews_max` dictionary created above to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews).

In [11]:
android_clean=[]
already_added=[]
    
for row in android:
    name = row[0]
    n_reviews=float(row[3])
    
    if n_reviews==reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)

#### 2.2.5 Inspect the `android_clean` dataset that we just created above

In [12]:
explore_data(android_clean,0,4,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 9659
Number of columns: 13


From the output of the previous cell, we confirmed that the `android_clean` dataset have 9,659 rows, just as expected, and that removing duplicates worked properly.

### 2.3 Removing Non-English Apps

When we explore the datasets enough, we notice that both datasets have apps with names that suggest they are not designed for an English-speaking audience:

In [13]:
print(ios[813][1])
print(ios[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


Since we use English for the apps we develop at our company, we'd like to analyze only the apps that are designed for an English-speaking audience, and remove the rests.

#### 2.3.1 Define a function to check whether a string contains a non-common English character

One way to do this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in `ord()` function to find out the corresponding encoding number of each character.

In [14]:
def english_character(string):
    
    for character in string:
        if ord(character)>127:
            return False
        
    return True

##### Trial: use the function to check the following app names

In [15]:
print(english_character('Instagram'))
print(english_character('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_character('Docs To Go™ Free Office Suite'))
print(english_character('Instachat 😜'))

True
False
False
False


As in the trial example above, the function coundn't correctly identify certain English App names that contain emojis or special characters like™. 

To minimize the impact of data loss, we modify the function to remove an app only if its name has more than threee non-ASCII characters.

In [16]:
def english_character(string):
    i=0
    
    for character in string:
        if ord(character)>127:
            i+=1
            
    if i>=3:
        return False
    else:
        return True

##### Trial: use the function to check the following app names

In [17]:
print(english_character('Instagram'))
print(english_character('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_character('Docs To Go™ Free Office Suite'))
print(english_character('Instachat 😜'))

True
False
True
True


Although the function is still not perfect, and very few non-English apps might still get past our filter, we'll take it as good enough at this point in our analysis, as in the above cell output. 

#### 2.3.2 Filter out non-English apps from both datasets

Next, we use the `english_character()` function to filter out the non-English apps for both datasets:

In [18]:
ios_english=[]

for app in ios:
    name=app[1]
    if english_character(name):
        ios_english.append(app)

android_english=[]

for app in android_clean:
    name=app[0]
    if english_character(name):
        android_english.append(app)

explore_data(ios_english,0,4,True)
print('\n')
explore_data(android_english,0,4,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 6155
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketc

After filtering, we derive the two datasets `ios_english` and `android_english`, with 5707 remaining entries in the ios dataset, and 9117 remaining entries in the android dataset. 

### 2.4 Isolating the Free Apps

As we mentioned in the introduction, we only build apps that are free to download and install. Therefore, we need to isolate only the free apps for our analysis for the next step. 

In [19]:
### for the ios dataset ###

ios_free=[]

for app in ios_english:
    price=app[4]
    if price=='0.0':
        ios_free.append(app)
    
explore_data(ios_free,0,10,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


['429047995', 'Pinterest', '74778624', 'USD', '0.0', '1061624', '1814', '4.5', '4.0', '6.26', '12+', 'Social Networking', '37', '5', '27', '1']


['282935706', 'Bible', '92774400', 'USD', '0.0', '985920', '5320', '4.5', '5.0', '7.5.1', '4+', 'Reference', '37', '5', '45', '1']


['5538347

In [20]:
### for the android dataset ###

android_free=[]

for app in android_english:
    price=app[7]
    if price=='0':
        android_free.append(app)
    
explore_data(android_free,0,10,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+

After removing the paid apps, we are left with the `ios_free` dataset with 2922 entries, and the `android_free` dataset with 9488 entries.

## 3. Determining Most Common Apps by Genre

As we mentioned in the introduction, our goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To minimize risks and overhead, our validation strategy for an app idea has three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. 

We begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the `prime_genre` column of the App Store data set, and the `Genres` and `Category` columns of the Google Play data set.

### 3.1 Build a Frequency Table of Genres

We'll build two functions we can use to analyze the frequency tables:
- One function `freq_table()` to generate frequency tables that show percentages
- Another function `display_table()` that we can use to display the percentages in a descending order

In [21]:
def freq_table(dataset,index):
    column=[]
    for app in dataset:
        column.append(app[index])

    dic={}
    total=0
   
    for item in column:
        total+=1
        if item in dic:
            dic[item]+=1
        else:
            dic[item]=1
    
    dic_percentage={}
    for item in column:
        dic_percentage[item]=dic[item]/total
    return dic_percentage


In [22]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Next, we start by examining the frequency table for the prime_genre column of the App Store data set:

In [23]:
display_table(ios_free, 11)

Games : 0.5825788323446769
Entertainment : 0.07836403371838901
Photo & Video : 0.049953168904152356
Education : 0.036840462066812366
Social Networking : 0.033093974399000935
Shopping : 0.025913206369029034
Utilities : 0.024664377146425227
Sports : 0.021542304089915705
Music : 0.020605682172962846
Health & Fitness : 0.020293474867311895
Productivity : 0.017483609116453323
Lifestyle : 0.015610365282547611
News : 0.013424914142990947
Travel : 0.012488292226038089
Finance : 0.010927255697783328
Weather : 0.008741804558226662
Food & Drink : 0.008117389946924758
Reference : 0.005307524196066188
Business : 0.005307524196066188
Book : 0.003746487667811427
Navigation : 0.0018732438339057135
Medical : 0.0018732438339057135
Catalogs : 0.001248829222603809


#### Comment:

From the output of the above cell, we can see that among all free English apps, more than half (58.26%) are `Games`. The next most common genre is `Entertainment`(7.83%), followed by `Photo & Video` apps (5.00%).


In general, we have the observation that **App Store is basically dominated by apps that are designed for fun** (games, entertainment, photos & videos, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. 

However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Next, we continue to explore the `Genres` and `Category` columns of the Google Play dataset (two columns which seem to be related):

In [24]:
### frequency table for Genres in android dataset ###
display_table(android_free,9)

Tools : 0.08442585895117541
Entertainment : 0.06080470162748644
Education : 0.05357142857142857
Business : 0.04599909584086799
Productivity : 0.03899186256781193
Lifestyle : 0.038765822784810125
Finance : 0.037070524412296565
Medical : 0.035375226039783
Sports : 0.03458408679927667
Personalization : 0.03322784810126582
Communication : 0.032323688969258586
Action : 0.03096745027124774
Health & Fitness : 0.030854430379746837
Photography : 0.029498191681735987
News & Magazines : 0.02802893309222423
Social : 0.02667269439421338
Travel & Local : 0.02328209764918626
Shopping : 0.02249095840867993
Books & Reference : 0.021360759493670885
Simulation : 0.020456600361663652
Dating : 0.018648282097649186
Arcade : 0.01842224231464738
Video Players & Editors : 0.017744122965641953
Casual : 0.01763110307414105
Maps & Navigation : 0.013901446654611212
Food & Drink : 0.012432188065099457
Puzzle : 0.011301989150090416
Racing : 0.009945750452079566
Role Playing : 0.009380650994575045
Libraries & Demo : 

In [25]:
### frequency table for Category in android dataset ###
display_table(android_free,1)

FAMILY : 0.18942133815551537
GAME : 0.09697106690777577
TOOLS : 0.08453887884267632
BUSINESS : 0.04599909584086799
PRODUCTIVITY : 0.03899186256781193
LIFESTYLE : 0.03887884267631103
FINANCE : 0.037070524412296565
MEDICAL : 0.035375226039783
SPORTS : 0.033905967450271246
PERSONALIZATION : 0.03322784810126582
COMMUNICATION : 0.032323688969258586
HEALTH_AND_FITNESS : 0.030854430379746837
PHOTOGRAPHY : 0.029498191681735987
NEWS_AND_MAGAZINES : 0.02802893309222423
SOCIAL : 0.02667269439421338
TRAVEL_AND_LOCAL : 0.02339511754068716
SHOPPING : 0.02249095840867993
BOOKS_AND_REFERENCE : 0.021360759493670885
DATING : 0.018648282097649186
VIDEO_PLAYERS : 0.017970162748643763
MAPS_AND_NAVIGATION : 0.013901446654611212
FOOD_AND_DRINK : 0.012432188065099457
EDUCATION : 0.01164104882459313
ENTERTAINMENT : 0.009606690777576853
LIBRARIES_AND_DEMO : 0.009380650994575045
AUTO_AND_VEHICLES : 0.009267631103074141
HOUSE_AND_HOME : 0.008024412296564195
WEATHER : 0.007911392405063292
EVENTS : 0.00712025316455

#### Comment:

From the output of the `Genres` column, the most common genre in the Google Play Store is `Tools` (8.44%), while the next most common is `Entertainment`(6.08%). 

From the output of the `Category` column, the most common category is `Family`(18.94%), followed by `Game`(9.70%) and `Tools`(8.45%).

It seems that things are a bit different for the Google Play store: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Still, practical apps seem to have a better representation on Google Play compared to App Store. 

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

### 3.2 Measuring Popularity –– the Average Number of Installs for each App

The average number of installs/ user ratings can be used as an approximation of the popularity of an app.

For the Google Play data set, we can find this information in the `Installs` column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` column.

#### For App Store:

Getting the frequency table of the `rating_count_tot` column:

In [26]:
ios_prime_genre=freq_table(ios_free,11)

Calculating the average `rating_count_tot` numbers per `Genres`:

In [27]:
average_user_ratings_dic={}

for genre in ios_prime_genre:
    total=0
    len_genre=0
    
    for item in ios_free:
        genre_app=item[11]
        
        if genre_app==genre:
            number_of_ratings=float(item[5])
            total+=number_of_ratings
            len_genre+=1
            
    average_user_ratings=total/len_genre
    average_user_ratings_dic[genre]=average_user_ratings
    print(genre,':',average_user_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22886.36709539121
Music : 57326.530303030304
Reference : 79350.4705882353
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 19156.493670886077
Travel : 28243.8
Shopping : 27230.734939759037
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16815.48
Entertainment : 14195.358565737051
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 46384.916666666664
Finance : 32367.02857142857
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Next, we create a function to sort the dictionary, so as to see which genre has the highest average number of user ratings, thus being the most popular:

In [28]:
def display_table_2(table):
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Sort the `average_user_ratings_dic` with the `display_table_2()` function just created above:

In [29]:
display_table_2(average_user_ratings_dic)

Navigation : 86090.33333333333
Reference : 79350.4705882353
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 46384.916666666664
Food & Drink : 33333.92307692308
Finance : 32367.02857142857
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 27230.734939759037
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22886.36709539121
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 19156.493670886077
Lifestyle : 16815.48
Entertainment : 14195.358565737051
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


From the output of the above cell, we can tell that in the App Store,  **Navigation** apps have the highest number of user reviews, with **Reference** and **Social Networking** following.

But taking a closer look, we find that this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [30]:
for app in ios_free:
    genre=app[11]
    if genre == 'Navigation':
        name=app[1]
        rating_count=app[5]
        print(name, ':', rating_count) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


It's similar for the `Reference` genre, where the Bible and Dictionary.com skewed up the average rating:

In [31]:
for app in ios_free:
    genre=app[11]
    if genre == 'Reference':
        name=app[1]
        rating_count=app[5]
        print(name, ':', rating_count) # print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
Jishokun-Japanese English Dictionary & Translator : 0


While we found out the several most popular genres, we also noticed that the average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold.

However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

#### For Google Play:

For the Google Play market, we have data about the number of `Installs` (with index 5), so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [32]:
display_table(android_free,5)

1,000,000+ : 0.1575497287522604
100,000+ : 0.11539330922242315
10,000,000+ : 0.10567359855334539
10,000+ : 0.10194394213381555
1,000+ : 0.0839737793851718
100+ : 0.06928119349005425
5,000,000+ : 0.06826401446654612
500,000+ : 0.05560578661844485
50,000+ : 0.04769439421338156
5,000+ : 0.04486889692585895
10+ : 0.035375226039783
500+ : 0.03243670886075949
50,000,000+ : 0.022830018083182642
100,000,000+ : 0.021360759493670885
50+ : 0.019213381555153706
5+ : 0.007911392405063292
1+ : 0.005085895117540687
500,000,000+ : 0.0027124773960217
1,000,000,000+ : 0.0022603978300180833
0+ : 0.0004520795660036166
0 : 0.00011301989150090416


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

##### We start by generating a frequency table for the `Category`  column (with index 1) of the Google Play dataset to get the unique app genres.

In [33]:
android_category=freq_table(android_free,1)

In [34]:
print(android_category)

{'ART_AND_DESIGN': 0.006442133815551537, 'AUTO_AND_VEHICLES': 0.009267631103074141, 'BEAUTY': 0.00599005424954792, 'BOOKS_AND_REFERENCE': 0.021360759493670885, 'BUSINESS': 0.04599909584086799, 'COMICS': 0.0061030741410488245, 'COMMUNICATION': 0.032323688969258586, 'DATING': 0.018648282097649186, 'EDUCATION': 0.01164104882459313, 'ENTERTAINMENT': 0.009606690777576853, 'EVENTS': 0.007120253164556962, 'FINANCE': 0.037070524412296565, 'FOOD_AND_DRINK': 0.012432188065099457, 'HEALTH_AND_FITNESS': 0.030854430379746837, 'HOUSE_AND_HOME': 0.008024412296564195, 'LIBRARIES_AND_DEMO': 0.009380650994575045, 'LIFESTYLE': 0.03887884267631103, 'GAME': 0.09697106690777577, 'FAMILY': 0.18942133815551537, 'MEDICAL': 0.035375226039783, 'SOCIAL': 0.02667269439421338, 'SHOPPING': 0.02249095840867993, 'PHOTOGRAPHY': 0.029498191681735987, 'SPORTS': 0.033905967450271246, 'TRAVEL_AND_LOCAL': 0.02339511754068716, 'TOOLS': 0.08453887884267632, 'PERSONALIZATION': 0.03322784810126582, 'PRODUCTIVITY': 0.03899186256

In [35]:
avg_install_dic={}

for category in android_category:
    total=0
    len_category=0
    
    for item in android_free:
        category_app=item[1]
        
        if category_app==category:
            number_install=item[5]
            number_install=number_install.replace('+','')
            number_install=number_install.replace(',','')
            
            total+=float(number_install)
            len_category+=1
    
    avg_install=total / len_category
    avg_install_dic[category]=avg_install
    print(category,':', avg_install)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8814199.78835979
BUSINESS : 1712290.1474201474
COMICS : 832613.8888888889
COMMUNICATION : 38590581.08741259
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1360598.042253521
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1446158.2238372094
GAME : 15544014.51048951
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3650602.276666667
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10830251.970588235
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5145550.285714285
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_MAGAZ

In [36]:
display_table_2(avg_install_dic)

COMMUNICATION : 38590581.08741259
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15544014.51048951
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10830251.970588235
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8814199.78835979
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5145550.285714285
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4049274.6341463416
FAMILY : 3695641.8198090694
SPORTS : 3650602.276666667
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1446158.2238372094
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1360598.042253521
DATING : 854028.8303030303
COMICS : 832613.8888888889
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 513151.886

On average, communication apps have the most installs: 38,590,581. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs: 

In [37]:
for app in android_free:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,814,159. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [38]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

It seems there's still a small number of extremely popular apps that skew the average:

In [39]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. 

Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [40]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.